### PR TITLE
easier build

### PR DESCRIPTION
**Pull request by arthurnn**
I worked on this because I finished a brew formula to install Vitess on macOS with one command: `brew install vitess`.
Also, there was some conversation on the Slack channel to improve the build process and make it easier.

The main change in here is: **we will push the vendor libs, instead of getting them during the bootstrap.sh script run.**
Advantages of this approach:
- We ensure the version of our libs, not just by the vendor.json file, but they actually will follow the history of this repo.
- Makes the build process a lot easier. With this, I can build Vitess without needing to run boostrap.sh.  This is useful for folks that don't need to change the code but just want to build it.
- Homebrew build process doesn't allow us to do `go get` during it, as that is insecure. But, as we vendor our dependencies we don't need to `go get` during build time anymore.

Drawbacks:
- when changing the vendor.json file, we need to run `go vendor sync` or the bootstrap script again, to upload the new version of the lib.
- This repository will get bigger in git size. (the local size will still be the same, though)

I think the advantages outweighs the disadvantages, as the ratio of local builds vs changing vendor.json file is bigger.

At, least at GitHub, thats what we do, we vendored and upload the lib sources to git. Also, open source projects such as Consul do the same.

Thoughts? 
cc @sougou @vmg

### CODE REVIEW

**Comment by guidoiaquinti** [comment]
+1 

minor: we should probably exclude these lib from `codeclimate` CI

**Comment by enisoc** [comment]
We used to check in vendored files like this, but then deleted them in #1636. Unfortunately the author of that PR (who has since left the team) neglected to explain the justification, so we can only try to guess.

One concern I have is that we need to make sure our copies don't diverge from upstream. For example, we need to mechanically block any PR that touches those files, unless those files match the version we depend on from the upstream repo. Otherwise, we might unknowingly take on patches to upstream code (which I believe has happened before).

In particular, I could see that being used to compromise Vitess binaries: if we don't check the hash of vendored files, one could sneak in changes in an "update dependencies" commit that no human will ever review (this PR has 3.8 million lines, for example).

It seems reasonable that a tool already exists to do this, so I'm not suggesting it's a deal-breaker for checking in these files, but I think it should be a prerequisite. In fact, govendor itself might be able to do this if we run `govendor status` as part of our travis build. We just need to confirm what it guarantees.

Other than that, I think the only reason we removed vendored files was for aesthetic reasons. If there are practical benefits to bringing them back, I'm ok with it.

**Comment by arthurnn** [comment]
Would love to get this through.  🙏 
Mainly because this improves the build process and time.
I believe we can only build vitess using Homebrew if we go ahead with this. Otherwise waiting for govendor sync during a `brew install vitess` will be a blocker because it takes a long time.

See bellow the time to install vitess with and without this using `brew install vitess`
```
### WIthout this PR
==> Summary
🍺  /usr/local/Cellar/vitess/2.1.1.08d7b71: 143 files, 257.4MB, built in 11 minutes 27 seconds

### With this PR
==> Summary
🍺  /usr/local/Cellar/vitess/2.1.1.75b5eb418f2b5d5ac1e8f5f8fd6bad723d0dcd48: 143 files, 257.4MB, built in 1 minute 27 seconds
```

> One concern I have is that we need to make sure our copies don't diverge from upstream. For example, we need to mechanically block any PR that touches those files, unless those files match the version we depend on from the upstream repo. Otherwise, we might unknowingly take on patches to upstream code (which I believe has happened before).

I think that is easy to solve with properly reviewing PRs, which I think is well handled in here. if there is any change in the vendor folder, that also don't bump the version in the vendor.json file, that should not be merged.


There is an alternative to this:
If we've done more frequent releases, and post them to https://github.com/youtube/vitess/releases as tar.gz file, and on that file, we include the vendor libs, then Homebrew could download that instead of the git source code.
Thoughts on that @sougou?

**Comment by ilovezfs** [comment]
Any reason not to use `dep` https://github.com/golang/dep ?

**Comment by derekperkins** [comment]
@ilovezfs govendor was picked before dep, and now that dep's future is in question, we're waiting to see how the vgo experiment turns out.

**Comment by ilovezfs** [comment]
@derekperkins ah ok. I've asked what the deal is with their outdated roadmap here: https://github.com/golang/dep/issues/1718

**Comment by sougou** [comment]
I'm still waiting for @alainjobart, @demmer and hopefully a few others to chime in.

For the release approach, an official release has compatibility rules and comes with paperwork of announcements, release notes, etc. If there is a way to make unofficial releases, then we could consider that approach.

My main concern comes from how big this PR is (3.8M lines), and it's a preview of what is to come. This will skew the dev stats of the project. Many people and tools, including CNCF, track these and use them to evaluate progress. So, there's value in trying to preserve the accuracy of these stats.

One work-around I thought of is to create a separate govendor repo and depend on that instead, which will address the above concern.

**Comment by demmer** [comment]
From my standpoint, all the options are distasteful in one way or another, but using a separate repository to check in the entire contents of the vendor dir and referencing that using a git submodule feels like the ~best~ least bad option.

I agree with all of @arthurnn's motivations for this change and the `brew install vitess` is a very tantalizing proposal.

But I also agree that million line PRs to change vendored dependencies are aesthetically unpleasant and unwieldy. Suppose we want to pull in a non-backwards-compatible version of an upstream package. If we were to check in the vendor code, then by definition a given PR needs to change both the vendor code and the vitess code that references it. With the submodule proposal, the vendor change would go into the separate vendor repo, then the vitess repository change would include the changed vitess bits and the updated git submodule reference.

The statistics question that sougou brought up is curious -- I would assume that CNCF's tooling to check for contributions could ignore changes to the vendor repo (at least if they don't then k8s would have the same challenges). But it's true that a casual observer of the project statistics would be skewed by large vendor changes.

So all in all, I'm in favor of either keeping things as they are and looking into some other ways to solve the homebrew problem, or looking into the submodule route.

**Comment by enisoc** [comment]
@arthurnn wrote:
> I think that is easy to solve with properly reviewing PRs, which I think is well handled in here. if there is any change in the vendor folder, that also don't bump the version in the vendor.json file, that should not be merged.

My request is that we have a solution that's more human-proof. Reviewing works if the reviewer always remembers the rule. Humans would also miss the malicious (or even accidental) case where the files being checked in don't match the promise made by the vendor.json record. Adding an automated check in Travis should be pretty straightforward if the tools already support such a command, and it would guard against both malicious and accidental divergence.

@ilovezfs wrote:
> Any reason not to use dep https://github.com/golang/dep ?

If we were starting now, I'd suggest dep over govendor since I've been using dep lately and like it. At this point though, I'm not aware of any benefit to switching for the specific case of Vitess. For example, I don't think we've hit the issue of vendored components that need to agree on the version of a common dependency. For the purpose of this PR, I don't think replacing `govendor sync` with `dep ensure` would make the build any easier. Both tools support checking in the vendored files as is being proposed here.

If anyone can think of benefits to switching, I'd love to hear them. It's probably orthogonal to this discussion though.

@sougou wrote:
> If there is a way to make unofficial releases, then we could consider that approach.

That might get around the Homebrew repo's technical requirement, but I think it would still violate the spirit of the rule. They only want to accept real, stable releases. The problem is we don't even have such a thing except in name; master is actually safer than any point-in-time release because we don't backport fixes.

@sougou wrote:
> My main concern comes from how big this PR is (3.8M lines)

I quoted that earlier too, based on GitHub's PR summary, but now I'm thinking it might be a glitch. I checked out this PR and ran `find vendor | xargs wc -l` and got a total of 388,893 lines, which seems more reasonable. I'm also only counting 1,212 files in `vendor/` (including directories), so I don't know where the rest of the 6,921 files changed are coming from.

@ilovezfs wrote:
> Another option is to use git submodules for everything in the vendor directory. That works nicely with caching too.

My concern with submodules is that we would effectively be using git to do the same thing as `govendor sync`, except that you could put everything in one `vitessio/vendor` repo so it's faster to download. Is the slow speed of `govendor sync` really the only problem we care about solving? Or would submodules have some additional benefit to the build process that I'm not aware of?

**Comment by demmer** [comment]
> Or would submodules have some additional benefit to the build process that I'm not aware of?

Just pointing out that submodules would help in the case where vitess needs to change alongside an upstream change. You can't have a blanket rule in travis which prevents the same PR from changing both vendored code and vitess code since you would then have no way of importing backwards-incompatible changes.

**Comment by enisoc** [comment]
@demmer wrote:
> You can't have a blanket rule in travis which prevents the same PR from changing both vendored code and vitess code

I wouldn't recommend such a rule anyway. It seems natural to me that changing vendored code would often require atomic changes to Vitess code, so these changes should be in the same PR (though perhaps different commits) so that master is always functional.

The rule I prefer is simply to always check the invariant that the contents of the vendor directory exactly match the hashes stored in the vendor.json file. This would prevent accidental acceptance of local patches to upstream, as well as preventing malicious code hidden in local patches of vendored code.

**Comment by demmer** [comment]
> I wouldn't recommend such a rule anyway. It seems natural to me that changing vendored code would often require atomic changes to Vitess code, so these changes should be in the same PR (though perhaps different commits) so that master is always functional.

Right. So it would mean that we will end up having to review a PR that contains both potentially large amounts of upstream vendor code changes as well as some vitess code. Separate commits would be a reasonable way to keep the review sane, but it's still a consideration.

**Comment by enisoc** [comment]
I think I misunderstood your earlier concern. I thought you were worried about making humans look at the vendored code, but it seems we agree that won't be necessary.

Now I'm thinking you were worried that the vendored code being included in the PR will make it more annoying to review the non-vendored code. If that's the case, I agree it would increase annoyance, and would only be partially mitigated by instituting a policy of putting vendored changes in their own PR.

If that's a correct interpretation of your concern, I think I should clarify what benefits I was asking about: I was asking for benefits of submodules over `govendor sync` (status quo), not over checking in vendored code (the proposal in this PR). For both submodules and `govendor sync`, PRs will remain free of vendored code.

**Comment by demmer** [comment]
> If that's a correct interpretation of your concern, I think I should clarify what benefits I was asking about: 

Yep that's the correct interpretation.

> I was asking for benefits of submodules over govendor sync (status quo), not over checking in vendored code (the proposal in this PR). For both submodules and govendor sync, PRs will remain free of vendored code.

@arthurnn should certainly weigh in here, but my understanding is:

(a) homebrew doesn't like running go get (but presumably can pull from git submodule?)
(b) govendor sync is "slow" and we think git is faster

**Comment by ilovezfs** [comment]
>(a) homebrew doesn't like running go get (but presumably can pull from git submodule?)

Correct. We don't allow`go get` at all. `git submodules` are allowed. Use git repository url + tag + revision, instead of tarball url + sha256, and the submodules will be cloned and cached.

>govendor sync is "slow" and we think git is faster

We do allow govendor sync since the git hashes are validated. Currently `govendor sync` is not able to take advantage of our caching mechanisms, whereas submodules are, though we could probably hack something together to get govendor sync to take advantage of caching if it comes to that.

Also, I'm not sure govendor and git submodules are mutually exclusive. Can't the submodules be maintained via govendor still?

**Comment by enisoc** [comment]
Since homebrew allows `govendor sync`, it seems our only problem with the status quo is that it's slow. Out of curiosity, I tried switching to dep and `dep ensure` took just as long.

> Currently govendor sync is not able to take advantage of our caching mechanisms, whereas submodules are, though we could probably hack something together to get govendor sync to take advantage of caching if it comes to that.

If it helps, govendor already looks for cached repos in `$GOPATH/.cache/govendor`. If the necessary repos are cached there, the time for `govendor sync` goes down from 76s to 3s on my machine.

> Also, I'm not sure govendor and git submodules are mutually exclusive. Can't the submodules be maintained via govendor still?

Yes, we can still use govendor either way. We're not considering replacing govendor with submodules - just replacing the `govendor sync` (download repos) step with submodules (or one big submodule).

**Comment by arthurnn** [comment]
Thanks all for the discussion.
So this aims to fix two things:
- Easier local build from source process
I guess we can get that in another way, we don't need to vendor everything. `govendor sync` is still totally acceptable in that case, especially because it should be fast as most people will have a warm cache locally.

- Faster build on Homebrew
Homebrew has no cache for govendor, that is why running a sync there is so slow. However, that is not the blocker, even if this gets merged, homebrew won't accept my PR. see Mike's comment:

> No, unfortunately this will need to be a stable, tagged release before we can include it into this Homebrew repository.

To summarize, merging this won't unblock us merging the homebrew side.

So, I would suggest we close this, however, in order to move forward with Homebrew, we would need to have a release cycle, and include the vendored files on the released .tar file.
I can write a script that creates the .tar for us, including the vendor libs. 
Does that make sense? We would probably need to release versions once or more per month, I would call them minor releases until we figure some good release and stable cycle.

The only thing is, I would not want to block the Homebrew side until we figure out a release cycle, if possible.

**Comment by enisoc** [comment]
Based on earlier comments from @sougou, it sounds like he's concerned that the current state of master does not fit the criteria for a v2.2.0 release according to semantic versioning.

I guess we need to decide if we want to change the way we develop Vitess to fit with semantic versioning, or abandon semantic versioning?

**Comment by shlomi-noach** [comment]
Chiming in to present my 2c.

> We ensure the version of our libs, not just by the vendor.json file, but they actually will follow the history of this repo.

I believe this is a critical criteria. I believe a project should be self-contained (or see a reasonable relaxation below), and without external dependencies.

> all the options are distasteful in one way or another

💯 

With `orchestrator` and `gh-ost` I use a completely distasteful approach of importing all dependencies onto my `vendor/` subdir, stripping out all `.git` info. I do not claim this to be the best approach, nor do I think it's the worst. I'm not suggesting it for Vitess.

`go get` during build is, in my opinion, an antipattern, and should be avoided. The project should be able to stand its own grounds, and I wish to avoid repeating all that was said above.

A reasonable exception/relaxation that I believe can be taken, if the owners don't want to bloat the `vitess` repo, is to create a single repo, `vitess-dep`, that will include all dependencies. This will keep `vitess` clean, will still require some `go get` or `git clone`, but only a single one, and will be fully owned by owners of `vitess`. We conclude that all options are distasteful and so is this one.

**Comment by shlomi-noach** [comment]
By the way, I'd argue that `govendor` itself, if used, needs to be vendored, or included in the `vitess` repo. But then again, it may not be needed in the first place. Infinite loop.

**Comment by sougou** [comment]
This has been lingering for a while now. I think we have two viable directions to take:
1. Move vendor stuff out into its own repo.
2. Make more frequent releases.

We didn't discuss details of option 2, but that is something we can also make work: We can immediately release a 3.0-alpha, and then we can commit to a release every two weeks using a minor version number.

When 3.0 is ready, we can release that, and continue the 'release every two weeks' cadence.

The reason why this is viable is because our master is pretty stable. So, there is no worry about releases being broken.

**Comment by ilovezfs** [comment]
Calling it "alpha" or "beta" or "rc" or "pre" will prevent its inclusion in Homebrew.

**Comment by sougou** [comment]
I was just informed that brew is mainly for mac (pardon my ignorance). Is this the case? If so, is this change meant to predominantly address a mac build?

**Comment by ilovezfs** [comment]
Homebrew is for macOS, correct. (The backend works on Linux too, though.)

**Comment by derekperkins** [comment]
👍 for bi-weekly builds.

Is the purpose for this to make it easier to onboard new devs who are going to contribute core Vitess code who are running on Mac? I'm not against making that easier, but it does seem like a tiny niche.

If I were new to Vitess, I guess I would rather have `brew install vitess` actually spin up a local cluster or something simple just to try it out. There are going to be a lot more users than contributors, so that seems like the path to smooth out. Maybe that is the intention here, so forgive me if I'm misunderstanding it. If that is the case, couldn't brew just use the release binaries rather than dealing with vendored code at all?

**Comment by bbeaudreault** [comment]
I think Option 2 would have other major benefits beyond homebrew, so probably makes most sense to just go with that. It does seem that homebrew would be used for ease-of-use for users, not contributors. From that standpoint it makes the most sense to base it off of a release, but our last release is incredibly outdated and would be hard to support. I think we should cut a true release ASAP, or at least define what the criteria will be for such a release.

**Comment by ilovezfs** [comment]
Building from source is required. By default people get the binary bottle built by CI, but they can pass `--build-from-source` or `--HEAD` if there is a head spec.

**Comment by sougou** [comment]
I kind of agree with @bbeaudreault. It's still worth making option 2 work for a number of other reasons:

* We haven't been releasing often. But it's a `bad thing`. We should release more often. Let's work on getting a process around this. And get someone allocated for it.
* We're not really a semantic versioning system. We instead guarantee backward compatibility with the previous release. I'm ok with this being our standard. After all, that's what our users want.
* I do want `3.0` to be special. We want to make a big splash, and use the CNCF marketing engine for this. But we need to fix the documentation before that. We'll look at getting this done before 3.0.

To unlock @arthurnn, we can make an interim (2.8?) release. But this still needs someone to compile the release notes: look at all the commits and build a summary of all the features. Who wants to volunteer for this?

Thoughts? Objections?

**Comment by arthurnn** [comment]
I like option 2 too the better.
So, having a release that includes the libs it is.
I will close this.

@sougou regarding the 2.8 release, I will try to put together some release notes.

Thanks all
